File: kotlin/samples/tictactoe/common/src/main/java/com/squareup/sample/gameworkflow/TakeTurnsWorkflow.kt
/*
 * Copyright 2017 Square Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.sample.gameworkflow

import com.squareup.sample.gameworkflow.Ending.Draw
import com.squareup.sample.gameworkflow.Ending.Quitted
import com.squareup.sample.gameworkflow.Ending.Victory
import com.squareup.sample.gameworkflow.RealTakeTurnsWorkflow.Action.Quit
import com.squareup.sample.gameworkflow.RealTakeTurnsWorkflow.Action.TakeSquare
import com.squareup.workflow.RenderContext
import com.squareup.workflow.Snapshot
import com.squareup.workflow.StatefulWorkflow
import com.squareup.workflow.Workflow
import com.squareup.workflow.WorkflowAction
import com.squareup.workflow.WorkflowAction.Updater

typealias TakeTurnsWorkflow = Workflow<TakeTurnsProps, CompletedGame, GamePlayScreen>

class TakeTurnsProps private constructor(
  val playerInfo: PlayerInfo,
  val initialTurn: Turn = Turn()
) {
  companion object {
    fun newGame(playerInfo: PlayerInfo): TakeTurnsProps = TakeTurnsProps(playerInfo)
    fun resumeGame(
      playerInfo: PlayerInfo,
      turn: Turn
    ): TakeTurnsProps = TakeTurnsProps(playerInfo, turn)
  }
}

/**
 * Models the turns of a Tic Tac Toe game, alternating between [Player.X]
 * and [Player.O]. Finishes with a [report][CompletedGame] of the last turn of the game,
 * and an [Ending] condition of [Victory], [Draw] or [Quitted].
 *
 * http://go/sf-taketurns
 */
class RealTakeTurnsWorkflow : TakeTurnsWorkflow,
    StatefulWorkflow<TakeTurnsProps, Turn, CompletedGame, GamePlayScreen>() {

  sealed class Action : WorkflowAction<Turn, CompletedGame> {
    class TakeSquare(
      private val row: Int,
      private val col: Int
    ) : Action() {
      override fun Updater<Turn, CompletedGame>.apply() {
        val newBoard = nextState.board.takeSquare(row, col, nextState.playing)

        when {
          newBoard.hasVictory() ->
            setOutput(CompletedGame(Victory, nextState.copy(board = newBoard)))

          newBoard.isFull() -> setOutput(CompletedGame(Draw, nextState.copy(board = newBoard)))

          else -> nextState = Turn(playing = nextState.playing.other, board = newBoard)
        }
      }
    }

    object Quit : Action() {
      override fun Updater<Turn, CompletedGame>.apply() {
        setOutput(CompletedGame(Quitted, nextState))
      }
    }
  }

  override fun initialState(
    props: TakeTurnsProps,
    snapshot: Snapshot?
  ): Turn = props.initialTurn

  override fun render(
    props: TakeTurnsProps,
    state: Turn,
    context: RenderContext<Turn, CompletedGame>
  ): GamePlayScreen = GamePlayScreen(
      playerInfo = props.playerInfo,
      gameState = state,
      onQuit = { context.actionSink.send(Quit) },
      onClick = { row, col -> context.actionSink.send(TakeSquare(row, col)) }
  )

  override fun snapshotState(state: Turn) = Snapshot.EMPTY
}

File: swift/Samples/TicTacToe/Sources/Game/TakeTurnsWorkflow.swift
/*
 * Copyright 2020 Square Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Workflow
import WorkflowUI

// MARK: Input and Output

struct TakeTurnsWorkflow: Workflow {
    var playerX: String
    var playerO: String

    typealias Output = Never
}

// MARK: State and Initialization

extension TakeTurnsWorkflow {
    struct State: Equatable {
        var board: Board
        var gameState: GameState
    }

    func makeInitialState() -> TakeTurnsWorkflow.State {
        return State(board: Board(), gameState: .ongoing(turn: .x))
    }

    func workflowDidChange(from previousWorkflow: TakeTurnsWorkflow, state: inout State) {}
}

// MARK: Actions

extension TakeTurnsWorkflow {
    enum Action: WorkflowAction {
        typealias WorkflowType = TakeTurnsWorkflow

        case selected(row: Int, col: Int)

        func apply(toState state: inout TakeTurnsWorkflow.State) -> TakeTurnsWorkflow.Output? {
            switch state.gameState {
            case let .ongoing(turn: turn):
                switch self {
                case let .selected(row: row, col: col):
                    if !state.board.isEmpty(row: row, col: col) {
                        return nil
                    }

                    state.board.takeSquare(row: row, col: col, player: turn)

                    if state.board.hasVictory() {
                        state.gameState = .win(turn)
                        return nil
                    } else if state.board.isFull() {
                        state.gameState = .tie
                        return nil
                    } else {
                        state.gameState.toggle()
                        return nil
                    }
                }

            case .tie:
                return nil
            case .win:
                return nil
            }
        }
    }
}

// MARK: Rendering

extension TakeTurnsWorkflow {
    typealias Rendering = GamePlayScreen

    func render(state: TakeTurnsWorkflow.State, context: RenderContext<TakeTurnsWorkflow>) -> Rendering {
        let sink = context.makeSink(of: Action.self)

        return GamePlayScreen(
            gameState: state.gameState,
            playerX: playerX,
            playerO: playerO,
            board: state.board.rows,
            onSelected: { row, col in
                sink.send(.selected(row: row, col: col))
            }
        )
    }
}
